crates/accelerate/src/sabre_layout.rs

```diff
@@ -96,39 +89,56 @@ pub fn sabre_layout_and_routing(
                     run_in_parallel,
                 )
             })
-            .min_by_key(|result| result.1.map.values().map(|x| x.len()).sum::<usize>())
+            .min_by_key(|(_, result)| result.map.map.values().map(|x| x.len()).sum::<usize>())
             .unwrap()
-    };
-    (result.0, result.1, result.2.into_pyarray(py).into())
+    }
 }
 
 fn layout_trial(
-    num_clbits: usize,
-    dag_nodes: &mut Vec<(usize, Vec<usize>, HashSet<usize>)>,
+    dag: &SabreDAG,
     neighbor_table: &NeighborTable,
     distance_matrix: &ArrayView2<f64>,
     heuristic: &Heuristic,
     seed: u64,
     max_iterations: usize,
     num_swap_trials: usize,
     run_swap_in_parallel: bool,
-) -> ([NLayout; 2], SwapMap, Vec<usize>) {
+) -> ([NLayout; 2], SabreResult) {
     // Pick a random initial layout and fully populate ancillas in that layout too
     let num_physical_qubits = distance_matrix.shape()[0];
     let mut rng = Pcg64Mcg::seed_from_u64(seed);
     let mut physical_qubits: Vec<usize> = (0..num_physical_qubits).collect();
     physical_qubits.shuffle(&mut rng);
     let mut initial_layout = NLayout::from_logical_to_physical(physical_qubits);
-    let mut rev_dag_nodes: Vec<(usize, Vec<usize>, HashSet<usize>)> =
-        dag_nodes.iter().rev().cloned().collect();
+    let new_dag_fn = |nodes| {
+        // Because the current implementation of Sabre swap doesn't permute
+        // the layout when placing control flow ops, there's no need to
+        // recurse into blocks. We remove them here, but still map control
+        // flow node IDs to an empty block list so Sabre treats these ops
+        // as control flow nodes, but doesn't route their blocks.
+        let node_blocks_empty = dag
+            .node_blocks
+            .iter()
+            .map(|(node_index, _)| (*node_index, Vec::with_capacity(0)));
+        SabreDAG::new(
+            dag.num_qubits,
+            dag.num_clbits,
+            nodes,
+            node_blocks_empty.collect(),
+        )
+        .unwrap()
+    };
+
+    let mut dag_forward: SabreDAG = new_dag_fn(dag.nodes.clone());
```

Comment:
This feels a bit weird to me, it feels like we're making an extra copy for the forward direction I guess this is to strip on the node blocks. but we want to retain that for the routing run later?

Reply:
Added a comment to make this clearer in 6d2ddd0.